### PR TITLE
Fix issue with status code ranges and shared routes

### LINKED ISSUE #2575: [OpenAPI3] `duplicate-header` error issued for shared routes when unioned ErrorModel contains a header
See [Playground](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvcmVzdCI7DQrSGmh0dHDWGnZlcnNpb25pbmfEIA0KdXNpbmcgVHlwZVNwZWMuSHR0cDvRFlJlc3TSFlbJS8VKQHVzZUF1dGgoDQogIEFwaUtlecQPPMYLTG9jYXRpb24uaGVhZGVyLCAiYXBpLWtleSI%2BIHwgT8QqMsUvW8Q%2FICB7xgcgIOQAujrHIUZsb3fEfy5pbXBsaWNpdCzIJmF1dGhvcml6xWRVcmw6ICLkAQRzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbS9jb21tb24vb8Q9Mi92Mi4wL8hIZSLJW3Njb3BlczogW8lSY29nbml0aXZlc2VydmljZXMuYXp1csZULmRlZmF1bHQiXcdBfcQHXT4NCikNCkDHMSjlAN50aXRsZTogIkHEPSBPcGVuQUkgQVBJxHt9yC5lcuUBUCJ7ZW5kcG9pbnR9L29wZW5hacQmINI9cyBmb3IgY29tcGxl5AEMcyBhbmQgc2VhcmNoxjPFfCAgyFA6IHN0cmluZ8UcfeUAp25hbWVzcGFjZSDFXi7GXuYB5WVycm9yDQptb2RlbCBFxA1SZXNwb25zZcZZQOYB4CAieC1tcy3FLi1jb2RlIshoxUcgIGRlc2NyaXDkAKPLGn3FZO0CeC5yZXNvdXJjZSgiZGVwbG95bWVudHMiKegAhETJFecAgXZpc2liaWxpdHkoInJlYWTEKyAgQGtlecZ4yC9JZNB5xlFBdWRpb1RyYW7pAKFPxQZzIHt92iTFHQ0KQGFjxA5TZXBhcmF0b3IoIi%2FkAIDHFygiYcQ5L3TMOuUAzEBzaGFyZWRSb3V0ZQ0Kb3AgZ2V00mVBc1BsYWluVGV4dOUCFuoAzTrrAQjlAcpvxlw%2FOvoA0CnoAQAgfO4ByuYB6v8Ayv8Ayv8AykFzyG1PYmplY%2F8Az%2F8Az8YcKdQc8wDb&e=%40typespec%2Fopenapi3&options=%7B%7D)

This previously worked in 0.48 and produced the following:
https://github.com/Azure/azure-rest-api-specs/blob/f4f14a07407eb37c8ff3fc1395736724130922be/specification/cognitiveservices/data-plane/AzureOpenAI/inference/2023-09-01-preview/generated.yaml#L34

```
 responses:
        '200':
          description: The request has succeeded.
          content:
            application/json:
              schema:
                anyOf:
                  - type: string
                  - $ref: '#/components/schemas/AudioTranscription'
```

The location of the error is also quite strange.

### PR DESCRIPTION
**Pull request by tjprescott**
Fix #2575.

### CODE REVIEW

**Review by timotheeguerin — APPROVED** [review]

**Review by timotheeguerin — COMMENTED** [review]
If this needs to be a hotfix, need to change the target branch

File: packages/openapi3/src/openapi.ts [hunk 1]
```diff
@@ -886,24 +886,33 @@ function createOAPIEmitter(program: Program, options: ResolvedOpenAPI3EmitterOpt
...
-          obj.headers[key] = getResponseHeader(value);
+          obj.headers[key] = headerVal;
         }
       }
     }
   }
 
+  // compare the JSON representation of the headers, ignoring order, and ensure they are equivalent
+  function areHeadersEquivalent(a: any, b: any): boolean {
+    const aJson = JSON.stringify(a, Object.keys(a).sort());
```
Comment (timotheeguerin) [inline packages/openapi3/src/openapi.ts:911]: This i think would be an issue if the header is not just a flat object. The sorting of keys is only done at the root here.

But maybe we can revisit when we have an issue

Comment (tjprescott) [reply packages/openapi3/src/openapi.ts:911]: I updated the logic to utilize `isDeepStrictEqual` which should handle nesting.

**Comment by tjprescott** [comment]
[Playground with Fix](https://cadlplayground.z22.web.core.windows.net/prs/2577/?c=aW1wb3J0ICJAdHlwZXNwZWMvcmVzdCI7DQrSGmh0dHDWGnZlcnNpb25pbmfEIA0KdXNpbmcgVHlwZVNwZWMuSHR0cDvRFlJlc3TSFlbJS8VKQHVzZUF1dGgoDQogIEFwaUtlecQPPMYLTG9jYXRpb24uaGVhZGVyLCAiYXBpLWtleSI%2BIHwgT8QqMsUvW8Q%2FICB7xgcgIOQAujrHIUZsb3fEfy5pbXBsaWNpdCzIJmF1dGhvcml6xWRVcmw6ICLkAQRzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbS9jb21tb24vb8Q9Mi92Mi4wL8hIZSLJW3Njb3BlczogW8lSY29nbml0aXZlc2VydmljZXMuYXp1csZULmRlZmF1bHQiXcdBfcQHXT4NCikNCkDHMSjlAN50aXRsZTogIkHEPSBPcGVuQUkgQVBJxHt9yC5lcuUBUCJ7ZW5kcG9pbnR9L29wZW5hacQmINI9cyBmb3IgY29tcGxl5AEMcyBhbmQgc2VhcmNoxjPFfCAgyFA6IHN0cmluZ8UcfeUAp25hbWVzcGFjZSDFXi7GXuYB5WVycm9yDQptb2RlbCBFxA1SZXNwb25zZcZZQOYB4CAieC1tcy3FLi1jb2RlIshoxUcgIGRlc2NyaXDkAKPLGn3FZO0CeC5yZXNvdXJjZSgiZGVwbG95bWVudHMiKegAhETJFecAgXZpc2liaWxpdHkoInJlYWTEKyAgQGtlecZ4yC9JZNB5xlFBdWRpb1RyYW7pAKFPxQZzIHt92iTFHQ0KQGFjxA5TZXBhcmF0b3IoIi%2FkAIDHFygiYcQ5L3TMOuUAzEBzaGFyZWRSb3V0ZQ0Kb3AgZ2V00mVBc1BsYWluVGV4dOUCFuoAzTrrAQjlAcpvxlw%2FOvoA0CnoAQAgfO4ByuYB6v8Ayv8Ayv8AykFzyG1PYmplY%2F8Az%2F8Az8YcKdQc8wDb&e=%40typespec%2Fopenapi3&options=%7B%7D)

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**